apps/dash-3d-image-partitioning/app.py

```diff
@@ -0,0 +1,935 @@
+                                        "background": "grey",
+                                    },
+                                ),
+                                # This store has to be put here so dcc.Loading sees that it is updating.
+                                dcc.Store(id="found-segs", data=found_seg_slices),
+                            ],
+                        ),
+                        html.Div(
+                            id="3D-graphs",
+                            children=[
+                                dcc.Graph(
+                                    "image-display-graph-3d",
+                                    figure=make_default_3d_fig(),
+                                    config=dict(displayModeBar=False,),
+                                )
+                            ],
+                            style={"display": "none"},
+                        ),
+                    ],
+                ),
+            ],
+        ),
+        dcc.Store(id="fig-3d-scene", data=default_3d_layout,),
+        dcc.Store(id="current-render-id", data=0),
+        dcc.Store(id="last-render-id", data=0),
+    ],
+)
+
+app.clientside_callback(
+    """
+function (show_seg_n_clicks) {
+    // update show segmentation button
+    var show_seg_button = document.getElementById("show-seg-check");
+    if (show_seg_button) {
+        show_seg_button.textContent = show_seg_n_clicks % 2 ?
+            "Show Segmentation" :
+            "Hide Segmentation";
+    }
+    var ret = (show_seg_n_clicks % 2) ? "" : "show";
+    return [ret,ret];
+}
+""",
+    [Output("show-hide-seg-2d", "children"), Output("show-hide-seg-3d", "children")],
+    [Input("show-seg-check", "n_clicks")],
+)
+
+app.clientside_callback(
+    """
+function(
+    image_select_top_value,
+    image_select_side_value,
+    show_hide_seg_2d,
+    found_segs_data,
+    image_slices_data,
+    image_display_top_figure,
+    image_display_side_figure,
+    seg_slices_data,
+    drawn_shapes_data) {{
+    let show_seg_check = show_hide_seg_2d;
+    let image_display_figures_ = figure_display_update(
+        [image_select_top_value,image_select_side_value],
+        show_seg_check,
+        found_segs_data,
+        image_slices_data,
+        [image_display_top_figure,image_display_side_figure],
+        seg_slices_data,
+        drawn_shapes_data),
+        // slider order reversed because the image slice number is shown on the
+        // other figure
+        side_figure = image_display_figures_[1],
+        top_figure = image_display_figures_[0],
+        d=3,
+        sizex, sizey;
+    // append shapes that show what slice the other figure is in
+    sizex = top_figure.layout.images[0].sizex,
+    sizey = top_figure.layout.images[0].sizey;
+    if (top_figure.layout.shapes) {{
+        top_figure.layout.shapes=top_figure.layout.shapes.concat([
+            tri_shape(d/2,sizey*image_select_side_value/found_segs_data[1].length,
```

Comment:
I don't understand what the `tri_shape` function does here, could you please explain?

Reply:
Ah yes, it returns a shape that looks like a triangle, I'll document it better.